### PR TITLE
Unify enum handling with generic utility function templates

### LINKED ISSUE #9043: Unify enum translations by using enumDisplayName
Some enums are translated to display strings in multiple locations and differently each time.
Port them to enumDisplayName and unify them.

See https://github.com/owncloud/client/blob/f29c794c5b4bca37190911cb2acd60b737498805/src/libsync/syncfileitem.h#L285 
https://github.com/owncloud/client/blob/f29c794c5b4bca37190911cb2acd60b737498805/src/libsync/syncfileitem.cpp#L78-L84

for an example.

### PR DESCRIPTION
**Pull request by fmoc**
Fixes #9043 among others.

Unfortunately, formatting of nested function calls and constructions of `QVariantMap`s has degraded. We may have to invest some time into adapting our clang-format config.

### CODE REVIEW

File: src/common/vfs.cpp [hunk 1]
```diff
@@ -259,3 +245,18 @@ const VfsPluginManager &VfsPluginManager::instance()
...
+template <>
+QString Utility::enumToString(Vfs::Mode mode)
+{
+    // Note: Strings are used for config and must be stable
+    switch (mode) {
+    case Vfs::Mode::WithSuffix:
+        return QStringLiteral("suffix");
+    case Vfs::Mode::WindowsCfApi:
+        return QStringLiteral("wincfapi");
+    default:
```
Comment (TheOneRing) [inline src/common/vfs.cpp:258]: never define default

File: src/gui/libcloudproviders/libcloudproviders.cpp [hunk 2]
```diff
@@ -12,8 +12,8 @@
  * for more details.
  */
 
-#include <cloudproviders/cloudprovidersaccountexporter.h>
-#include <cloudproviders/cloudprovidersproviderexporter.h>
+#include <cloudprovidersaccountexporter.h>
```
Comment (TheOneRing) [inline src/gui/libcloudproviders/libcloudproviders.cpp:15]: unrelated

File: src/libsync/syncresult.cpp [hunk 3]
```diff
@@ -34,43 +34,43 @@ void SyncResult::reset()
     *this = SyncResult();
 }
 
-QString SyncResult::statusString() const
+template <>
+QString Utility::enumToDisplayName(SyncResult::Status status)
```
Comment (TheOneRing) [inline src/libsync/syncresult.cpp:38]: https://github.com/owncloud/client/issues/10882

File: src/gui/libcloudproviders/libcloudproviders.cpp (hunk 2 above)
Comment (fmoc) [reply src/gui/libcloudproviders/libcloudproviders.cpp:15]: It did not want to build locally, so I thought I'd include it...

File: src/common/utility.h [hunk 4]
```diff
@@ -347,7 +347,7 @@ OCSYNC_EXPORT Q_DECLARE_LOGGING_CATEGORY(lcUtility)
     }
 
     template <class E>
-    QString enumToString(E value)
+    OCSYNC_EXPORT QString enumToString(E value)
```
Comment (TheOneRing) [inline src/common/utility.h:350]: One compile issue is that you added the export macro to an inline function

Comment (TheOneRing) [reply src/common/utility.h:350]: it was the only issue, and actually the compiler error was quite exact 😅

Comment (fmoc) [reply src/common/utility.h:350]:
```suggestion
    QString enumToString(E value)
```

File: src/common/checksumalgorithms.cpp [hunk 5]
```diff
@@ -38,4 +38,15 @@ CheckSums::Algorithm CheckSums::fromByteArray(const QByteArray &s)
     return fromName(s.toUpper().constData());
 }
 
+template <>
+QString Utility::enumToString(CheckSums::Algorithm algo)
+{
+    const auto n = toString(algo);
+#if (QT_VERSION < QT_VERSION_CHECK(6, 0, 0))
```
Comment (TheOneRing) [inline src/common/checksumalgorithms.cpp:45]: you can kill the ifdef

File: src/common/pinstate.cpp [hunk 6]
```diff
@@ -14,3 +14,24 @@
...
+{
+    switch (availability) {
+    case VfsItemAvailability::AlwaysLocal:
+        return QCoreApplication::translate("pinstate", "Always available locally");
+    case VfsItemAvailability::AllHydrated:
+        return QCoreApplication::translate("pinstate", "Currently available locally");
+    case VfsItemAvailability::Mixed:
+        return QCoreApplication::translate("pinstate", "Some available online only");
+    case VfsItemAvailability::AllDehydrated:
+    case VfsItemAvailability::OnlineOnly:
```
Comment (TheOneRing) [inline src/common/pinstate.cpp:33]: fallthrough

File: src/gui/libcloudproviders/libcloudproviders.cpp (hunk 2 above)
Comment (TheOneRing) [reply src/gui/libcloudproviders/libcloudproviders.cpp:15]: please fix

File: src/libsync/propagateuploadtus.cpp [hunk 7]
```diff
@@ -85,8 +85,8 @@ SimpleNetworkJob *PropagateUploadFileTUS::makeCreationWithUploadJob(QNetworkRequ
     auto addMetaData = [&encodedMetaData](const QByteArray &key, const QByteArray &value) { encodedMetaData << key + ' ' + value.toBase64(); };
     addMetaData(QByteArrayLiteral("filename"), propagator()->fullRemotePath(_item->_file).toUtf8());
     // in difference to the old protocol the algrithm and the value are space seperated
-    addMetaData(QByteArrayLiteral("checksum"), CheckSums::toQString(checksumHeader.type()).toUtf8() + ' ' + checksumHeader.checksum());
-    addMetaData(QByteArrayLiteral("mtime"), QByteArray::number(_item->_modtime));
+    addMetaData(QByteArrayLiteral("checksum"), Utility::enumToString(checksumHeader.type()).toUtf8() + ' ' + checksumHeader.checksum());
+    addMetaData(QByteArrayLiteral("mtime"), QByteArray::number(static_cast<int>(_item->_modtime)));
```
Comment (TheOneRing) [inline src/libsync/propagateuploadtus.cpp:89]: On unix this is a down cast from int64 to int32?

File: src/libsync/syncresult.cpp [hunk 8]
```diff
@@ -34,47 +34,34 @@ void SyncResult::reset()
...
+    case SyncResult::Status::Problem:
+        return QStringLiteral("Success, some files were ignored.");
+    case SyncResult::Status::SyncAbortRequested:
+        return QStringLiteral("Aborting sync");
+    case SyncResult::Status::Paused:
+        return QStringLiteral("Sync paused");
+    case SyncResult::Status::Offline:
+        return QStringLiteral("Offline");
     }
-    return re;
```
Comment (TheOneRing) [inline src/libsync/syncresult.cpp:77]: 😍

**Review by TheOneRing — CHANGES_REQUESTED** [review]

**Review by TheOneRing — APPROVED** [review]
